flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java

```diff
@@ -297,9 +318,279 @@ private static Object handleLogicalTypeSchema(Schema fieldSchema, Object value)
+        @VisibleForTesting
+        static String convertUUID(Object value) {
+            if (value instanceof UUID) {
+                return ((UUID) value).toString();
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof String, "Expecting a value as String type.");
+                UUID.fromString((String) value);
+                return (String) value;
+            }
+        }
+
+        private static long validateTimestamp(long timestamp) {
+            Timestamp minTs = Timestamp.parseTimestamp("0001-01-01T00:00:00.000000+00:00");
+            Timestamp maxTs = Timestamp.parseTimestamp("9999-12-31T23:59:59.999999+00:00");
+            try {
+                // Validate timestamp from microseconds since EPOCH.
+                Timestamp.ofTimeMicroseconds(timestamp);
+                return timestamp;
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Invalid Timestamp '%s' Provided."
+                                        + "\nShould be a long value indicating microseconds since Epoch (1970-01-01 00:00:00.000000+00:00) "
+                                        + "between %s and %s",
+                                timestamp, minTs, maxTs));
+            }
+        }
+
+        // BigQuery inputs timestamp as microseconds since EPOCH,
+        // So if we have TIMESTAMP in micros - we convert as it is.
+        // If the TIMESTAMP is in millis - we convert to Micros and then add.
+        @VisibleForTesting
+        static Long convertTimestamp(Object value, boolean micros, String type) {
+            long timestamp;
+            if (value instanceof ReadableInstant) {
+                timestamp = TimeUnit.MILLISECONDS.toMicros(((ReadableInstant) value).getMillis());
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof Long,
+                        String.format(
+                                "Expecting a value as Long type %s. Instead %s was obtained",
+                                type, value.getClass()));
+                timestamp = (micros ? (Long) value : TimeUnit.MILLISECONDS.toMicros((Long) value));
+            }
+            return validateTimestamp(timestamp);
+        }
+
+        private static Integer validateDate(Integer date) {
+            if (date > 2932896 || date < -719162) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Invalid Date '%s' Provided."
+                                        + "\nShould be a Integer value indicating days since Epoch (1970-01-01 00:00:00) "
+                                        + "between %s and %s",
+                                LocalDate.ofEpochDay(date), "0001-01-01", "9999-12-31"));
+            }
+            return date;
+        }
```

Comment:
Lets not have these methods return anything. A validating method should throw an exception or just return void.

Reply:
ACK, Fixed.